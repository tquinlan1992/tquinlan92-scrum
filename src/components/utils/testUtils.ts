import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { AnyAction, isType, ActionCreator } from 'typescript-fsa';
import { AppState } from '../../headless/store';

export function getAnyJestFn() {
    return jest.fn() as any;
}

interface AnyActionPayload extends AnyAction {
    payload: any;
}

export function expectActionWithPayload(actionToTest: AnyActionPayload, expectedActionCreator: ActionCreator<any>, expectedPayload?: any) {
    expect(isType(actionToTest, expectedActionCreator)).toBe(true);
    expect(actionToTest.payload).toEqual(expectedPayload);
}


const middlewares = [thunk];
export const mockStore = configureStore(middlewares);

export function getMockStore(state: Partial<AppState>) {
    return mockStore(state);
}

export function expectCalledOnceWith(mockFunction: jest.Mock<{}>, calledWith?: any) {
    expect(mockFunction).toHaveBeenCalledTimes(1);
    const expectedParams = calledWith ? [calledWith] : [];
    expect(mockFunction.mock.calls[0]).toEqual(expectedParams);
}

export function mockClearAll(mockFunctions: jest.Mock<{}>[]) {
    mockFunctions.forEach(mockFunction => {
        mockFunction.mockClear();
    })
}