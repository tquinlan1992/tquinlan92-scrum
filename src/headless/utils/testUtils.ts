export function mockPouchDB() {
    jest.mock('../database/core/pouch', () => {
        return {};
    });
}
