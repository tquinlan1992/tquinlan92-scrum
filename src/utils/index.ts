import * as _ from 'lodash';
import { connect, MapStateToPropsParam, MergeProps } from 'react-redux';
import { AppState } from '@headless/store';
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';

type PropsMap<M, K extends keyof M> = {[P in K]: M[P]};

export function pick<M extends object, K extends keyof M>(obj: M, ...props: K[]): PropsMap<M, K> {
    // @ts-ignore: TS2352: Type 'PartialDeep<M>' cannot be converted to type 'PropsMap<M, K>'.
    return _.pick<M>(obj, props) as PropsMap<M, K>;
}

export type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];

export type AppStateThunkDispatch = ThunkDispatch<AppState, undefined, AnyAction>;
