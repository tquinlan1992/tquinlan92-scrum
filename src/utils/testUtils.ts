export function mockPouchDB() {
    jest.mock('../headless/database/pouch', () => {
        return {};
    });
}

export function mockPouch(dbMethods: object) {
    jest.mock('../headless/database/pouch', () => {
        return {

        }
    })
    const mockPouchRequire = require('../headless/database/pouch');
    const mockGetRemoteDb = jest.fn(() => {
        return new Promise(resolve => {
            resolve(dbMethods);
        })
    });
    mockPouchRequire.getRemoteDb = mockGetRemoteDb;
    return mockPouchRequire;
}

export function mockShortId() {
    jest.mock('shortid', () => {
        let count = 0;
        return () => {
            count = count + 1;
            return String(count);
        }
    })
}

