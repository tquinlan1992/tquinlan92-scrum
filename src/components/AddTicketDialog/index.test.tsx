jest.mock('./redux', () => {
    return {
        actions: {}
    };
});
import * as React from 'react';
import { shallow, configure } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import { AddTicketDialog } from './';
import { getAnyJestFn } from '../utils/testUtils';
import * as _ from 'lodash';

configure({ adapter: new Adapter() });

const mockActions = {
    onRequestClose: getAnyJestFn(),
    addTicket: getAnyJestFn(),
    onSubmit: getAnyJestFn()
};

function testShowAddTicketDialogValue(open: boolean) {
    describe(`open as ${open}`, () => {
        it(`AddTicketDialog should show the addTicketDialog open property as ${open}`, () => {
            const props = {
                open,
                onRequestClose: mockActions.onRequestClose,
                addTicket: mockActions.addTicket as any,
                reset: jest.fn() as any,
                setAddTicketState: jest.fn() as any,
                onSubmit: mockActions.onSubmit,
                storyPoint: 5,
                description: 'description',
                title: 'title'
            };
            const result = shallow(<AddTicketDialog {...props} />);
            expect(result).toMatchSnapshot();
        });
    });
}

describe('describe when the AddTicketDialog Component is given', () => {
    testShowAddTicketDialogValue(true);
    testShowAddTicketDialogValue(false);
});

describe('when a user', () => {
    beforeEach(() => {
        _.forEach(mockActions, mockAction => {
            mockAction.mockClear();
        });
    });
    describe('requests to close addTicket', () => {
        it('setShowAddTicketDialog should be called with false', () => {
            const props = {
                open: true,
                onRequestClose: mockActions.onRequestClose,
                addTicket: mockActions.addTicket as any,
                reset: jest.fn() as any,
                setAddTicketState: jest.fn() as any,
                onSubmit: mockActions.onSubmit,
                storyPoint: 5,
                description: 'description',
                title: 'title'
            };
            const result = shallow(<AddTicketDialog {...props} />);
            const dialogProps: any = result.find('[title="Dialog With Actions"]').props();
            dialogProps.onExit();
            expect(mockActions.onRequestClose.mock.calls).toMatchObject([
                []
            ]);
        });
    });
});
