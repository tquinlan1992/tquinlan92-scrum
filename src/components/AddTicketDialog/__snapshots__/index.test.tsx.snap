// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`describe when the AddTicketDialog Component is given open as false AddTicketDialog should show the addTicketDialog open property as false 1`] = `
<WithStyles(Dialog)
  fullWidth={true}
  onExit={[Function]}
  open={false}
  title="Dialog With Actions"
>
  <WithStyles(DialogTitle)
    id="form-dialog-title"
  >
    Create a Ticket
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <form
      autoComplete="off"
    >
      <WithStyles(FormControl)>
        <TextField
          autoFocus={true}
          fullWidth={true}
          id="name"
          label="Title"
          margin="dense"
          onChange={[Function]}
          required={true}
          select={false}
          type="text"
          value="title"
          variant="standard"
        />
      </WithStyles(FormControl)>
      <WithStyles(FormControl)>
        <TextField
          fullWidth={true}
          id="multiline-flexible"
          label="Description"
          margin="dense"
          multiline={true}
          onChange={[Function]}
          required={false}
          select={false}
          value="description"
          variant="standard"
        />
      </WithStyles(FormControl)>
      <Connect(StoryPointsInputComponent) />
    </form>
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </WithStyles(Button)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Create
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;

exports[`describe when the AddTicketDialog Component is given open as true AddTicketDialog should show the addTicketDialog open property as true 1`] = `
<WithStyles(Dialog)
  fullWidth={true}
  onExit={[Function]}
  open={true}
  title="Dialog With Actions"
>
  <WithStyles(DialogTitle)
    id="form-dialog-title"
  >
    Create a Ticket
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <form
      autoComplete="off"
    >
      <WithStyles(FormControl)>
        <TextField
          autoFocus={true}
          fullWidth={true}
          id="name"
          label="Title"
          margin="dense"
          onChange={[Function]}
          required={true}
          select={false}
          type="text"
          value="title"
          variant="standard"
        />
      </WithStyles(FormControl)>
      <WithStyles(FormControl)>
        <TextField
          fullWidth={true}
          id="multiline-flexible"
          label="Description"
          margin="dense"
          multiline={true}
          onChange={[Function]}
          required={false}
          select={false}
          value="description"
          variant="standard"
        />
      </WithStyles(FormControl)>
      <Connect(StoryPointsInputComponent) />
    </form>
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </WithStyles(Button)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Create
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;
